### ISSUE TITLE
extauth: new timeout behavior does not add the grpc-timeout header

### ISSUE THREAD

**Issue by yuval-k** [envoyproxy/envoy#13580]
The new timeout behavior for ext authz introduced in https://github.com/envoyproxy/envoy/pull/12778 doesn't pass the timeout to the async client. as a result the client does not add the grpc-timeout header. this header is used by the usptream service to be aware of the client timeout.

Suggestion: we can add a deadline field to the request options, that if present, will be used for the calculation of the grpc-timeout header. note that deadline is used as time may pass between the creation of the request and until it is sent upstream.

see slack discussion: https://envoyproxy.slack.com/archives/C78M4KW76/p1602583831191600

cc @htuch

**Comment by htuch** [comment]
@yuval-k SGTM. At some point it might make sense to have both Envoy and Google gRPC take full responsibility for handling all aspects of the deadline, including being aware of queue delay, etc.

**Comment by yuval-k** [comment]
looking at the code a bit, its more complicated than i thought. Ideally we want to add the timeout header just before the request is sent upstream, here: https://github.com/envoyproxy/envoy/blob/eb9af96cd317b823ee6822dc10aeb26d63e51608/source/common/router/upstream_request.cc#L419

but grpc header is set in the async client ([here](https://github.com/envoyproxy/envoy/blob/e5aa69658c6182dd41b6217ec7f6c4c00cac84b4/source/common/grpc/async_client_impl.cc#L88)) and in the router filter ([here](https://github.com/envoyproxy/envoy/blob/eb9af96cd317b823ee6822dc10aeb26d63e51608/source/common/router/router.cc#L201)).

We can potentially set the grpc timeout in the UpstreamRequest::onPoolReady. This requires some thinking as the RequestOptions object is not propagated to the Router/UpstreamRequest, and I don't think we want to change the envoy timeout semantics globally. Any thoughts or feedback here are welcome.

One problem I see if I'm reading the code correctly, is due to re-try hedging, multiple upstream requests will modify the same headers map (to set the timeout). I believe (but did not test) that this is already an issue with auto host re-write. while this will not cause memory corruption, it might lead to incorrect values in the headers going upstream when using hedge retry - i.e. the host header set to all hedged upstreams will be the same.

**Comment by htuch** [comment]
Clarifying a few things as this has now become somewhat more complicated:
1. We originally had the issue that the timeout was not being respected when circuit breakers were enabled, since it was only set when a request was issued, so queue time didn't contribute to timeout.
2. https://github.com/envoyproxy/envoy/pull/12778 added support for enforcing a timeout in `ext_authz`, but doesn't set this on the gRPC async client. 
3. We've now lost `grpc-timeout`, which used to be set when configuring the gRPC async client.

This makes sense, but I'm also a bit confused. When I look at the Envoy gRPC async client, it looks like the timeout is [used](https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.cc#L90) for `grpc-timeout`, but not for actually setting a router timeout, so I'm not sure how (1) could have arisen in the first place. The router does [set](https://github.com/envoyproxy/envoy/blob/5c7737266e671ea9801c14d2779ca30bb0032bf7/source/common/router/router.cc#L120) the request timeout based on the `grpc-timeout` header, but only if `maxGrpcTimeout()` is non-null on the route, which it always is for `AsyncClient`. This is further compounded by the header computation for `grpc-timeout` happening prior to queueing on stream [initialization](https://github.com/envoyproxy/envoy/blob/e5aa69658c6182dd41b6217ec7f6c4c00cac84b4/source/common/grpc/async_client_impl.cc#L69).

I probably missed something, but want to understand exactly the relationship between setting the `grpc-timeout` header and timers before the change.

**Comment by yuval-k** [comment]
i believe the router also sets the timeout based on a timeout on the route, and the async client also implements a route for the router filter

**Comment by yuval-k** [comment]
@htuch until we figure the proper way to do this, should I disable turning on the new feature by default in the meantime?

**Comment by htuch** [comment]
Got it, thanks. Yeah, please flip the default for now.

Why do we want to set the header in `onPoolReady()`? If we set the `grpc-timeout` as it is done today, then the remote server sees roughly the same timeout as the one ext_authz is implementing.

One other thing occurred to me, should we also be applying the ext_authz timeout fix to ratelimit? Wondering if the right thing to do is make the gRPC Envoy client explicitly aware of the timeout. I think it's fine to just fix for grpc-timeout to resolve the immediate issue but would be good to understand the longer term fix.

**Comment by mattklein123** [comment]
> One other thing occurred to me, should we also be applying the ext_authz timeout fix to ratelimit? Wondering if the right thing to do is make the gRPC Envoy client explicitly aware of the timeout. I think it's fine to just fix for grpc-timeout to resolve the immediate issue but would be good to understand the longer term fix.

+1, I don't think I spoke up last time, but the previous PR made me pretty uncomfortable. IMO we should actually revert it and come up with a better and more holistic solution that is generic to AsyncClient.

**Comment by yuval-k** [comment]
If I understand the code correctly `onPoolReady()` happens after the is no longer pending, right? so sounds like the deadline to timeout math should happen there? You probably know better than me..

**Comment by yuval-k** [comment]
> +1, I don't think I spoke up last time, but the previous PR made me pretty uncomfortable. IMO we should actually revert it and come up with a better and more holistic solution that is generic to AsyncClient.

I will create a PR now to remove it from the enabled features; If it's ok i would like to have the option to enable this, so we can do some further internal testing.

**Comment by yuval-k** [comment]
> Why do we want to set the header in onPoolReady()? If we set the grpc-timeout as it is done today, then the remote server sees roughly the same timeout as the one ext_authz is implementing.

The flow as I understand it (please correct me here):
1. grpc async client creates an http async client, and passes it the RequestOptions
1. http async client creates a router filter and a route entry. The route entry has the timeout
1. http async client uses the router filter to perform the request
1. while decoding headers, the router filter creates a grpc timeout header based on the timeout in the route
1. the router filter creates an upstream request, and calls `UpstreamRequest::encodeHeaders`
1. upstream request calls `conn_pool_->newStream`, with a callback
1. If I understand this correctly, at this point circuit breaker delay can happen
1. once the request can be sent upstream, UpstreamRequest::onPoolReady is called, and the headers are encoded upstream

Assuming the above is correct, I think the desired behavior is to set the upstream timeout header to whatever time is left given the original grpc async client timeout, right?

**Comment by htuch** [comment]
> Assuming the above is correct, I think the desired behavior is to set the upstream timeout header to whatever time is left given the original grpc async client timeout, right?

Yes, but we could also set it earlier as already done today and we get the correct overall timeout. I don't think the math with `grpc-timeout` was the original issue, it was rather when the per-route timeout was activated; it happened (I think) around step 8, rather than at step 1 (prior to circuit breaker delay).

I think a flag flip is fine to start with, we can discuss whether to rollback or how to fix once that lands.

**Comment by yuval-k** [comment]
if we set the timeout header earlier (say in step 1), it might not reflect the real timeout when it sent upstream (in step 8).
i.e. if the time out is 1s, and the circuit breaker delay is 400ms, then the upstream will receive a header set to 1 second, where it should get 600ms instead.

**Comment by mattklein123** [comment]
Your description of how it works is correct. I would like to step back and better understand what timeout we are trying to achieve across the queue delay, and I *definitely* want to do this generically in the async client / router. I don't think it will be hard to plumb things through once we understand the desired behavior.

I'm OK with keeping the original patch temporarily, but I would rather you just apply it locally and we revert it.

**Comment by yuval-k** [comment]
> Your description of how it works is correct. I would like to step back and better understand what timeout we are trying to achieve across the queue delay, and I definitely want to do this generically in the async client / router. I don't think it will be hard to plumb things through once we understand the desired behavior.

I understand and agree.

As for defining the desired behavior: I think we all agree that timeouts in the async client should start when the async client request is created. 
I believe we also agree that this is true for ext auth and rate limit.  Is this true for regular http request as well? do we expect different requirements for different grpc/http requests?

As for the grpc timeout header, IMHO, it should be set just before the request goes upstream (i.e. in `UpstreamRequest::onPoolReady`). and it should be set to the amount of time left for request (i.e. minus any time the request spent in pending due to circuit breaking). This logic is a bit more complicated, as we need to figure-out the amount time left. does that make sense?

As for doing this generically - the see a challenge in figuring out how to do this for the async client, but without impacting for the regular request for when the router filter is used via the HCM, as I assume we want to keep the existing behavior the same there?

WDYT?

**Comment by mattklein123** [comment]
That all sounds correct to me. AsyncClient is just a thin wrapper around the router filter so I think we can pretty easily add whatever "configuration" we need to make it work.

> As for doing this generically - the see a challenge in figuring out how to do this for the async client, but without impacting for the regular request for when the router filter is used via the HCM, as I assume we want to keep the existing behavior the same there?

Per above we can have "config" that is different between the real router case and the async client case if we want, though I'm not sure we want to do that. If we fix things for the better I think we can do it generically as long as it's behind a runtime flag. Let's discuss in PR?

**Comment by yuval-k** [comment]
Sounds good - would the next step be a PR implementing the above?
I can start working on it shortly.

**Comment by mattklein123** [comment]
> Sounds good - would the next step be a PR implementing the above?
I can start working on it shortly.

Yes. Either revert the extauth change in the same PR or in a different PR. I'm fine either way.

**Comment by htuch** [comment]
> if we set the timeout header earlier (say in step 1), it might not reflect the real timeout when it sent upstream (in step 8).
i.e. if the time out is 1s, and the circuit breaker delay is 400ms, then the upstream will receive a header set to 1 second, where it should get 600ms instead.

Ah, right, because `grpc-timeout` is a relative duration, not a wall-clock timestamp (for some reason I was thinking the latter). Yep, in that case we need to set it when ready.

**Comment by yuval-k** [comment]
started working on ths

**Comment by mattklein123** [comment]
@yuval-k great, thanks. I actually had it on my list to revert the original PR so I will hold off on that. :)
